File: topics/type-decl/examples/shape_drawing.cpp
/* Program: Shape Drawer - SwinGame source file. (C++) */
#include "SwinGame.h"

// =====================
// = Declare Constants =
// =====================
#define MAX_SHAPES  20
#define MENU_RIGHT_X 60

// Menu Shape Constants
const rectangle MENU_RECT       = {10,250,40,50};
const circle    MENU_CIRCLE     = {{30,100},20};
const triangle  MENU_TRIANGLE   = {{30,150},{10,200},{50,200}};
const rectangle MENU_ELLIPSE    = {10,350,40,50};
const rectangle DRAWING_PAD     = {61,0,790,600};

// =================
// = Declare Types =
// =================

// The Shape Type is one of these options
typedef enum
{
    CIRCLE,
    RECTANGLE, //TODO: Add Ellipse and Triangle
    UNKNOWN
} shape_type;

// The Shape Data is either...
typedef union
{
    rectangle   rect;
    circle      circ; //TODO: Add Ellipse (as rectangle) and Triangle
} shape_data;

// The shape records...
typedef struct
{
    shape_type  type;        // the option selected
    color       fill_color;  // the fill color
    shape_data  data;        // the shape's data
} shape;

// A Drawing records...
typedef struct
{
    shape       shapes[MAX_SHAPES]; // a number of shapes
    int         index;              // the index of the last shape used
    shape_type  selected_shape;     // the selected shape type
} drawing;

File: topics/type-decl/examples/ShapeDrawing.pas
program ShapeDrawer;
uses sgTypes, sgInput, sgAudio, sgGraphics, sgResources, sgGeometry;
    
// =====================
// = Declare Constants =
// =====================
const MAX_SHAPES = 20;

const MENU_RECT     : Rectangle = (x: 10; y: 250; width: 40; height: 50);
const MENU_CIRCLE   : Circle = (center: (x: 30; y: 100); radius: 20);
const DRAWING_PAD   : Rectangle = (x: 61; y: 0; width: 790; height: 600);
// Todo: Add Ellipse & Triangle

// =================
// = Declare Types =
// =================
type
    // The Shape Type is one of these options
    // Todo: add Ellipse & Triangle
    ShapeType = ( CIRCLE_TYPE, RECTANGLE_TYPE, UNKNOWN_TYPE ); 
    
    Shape = record                      // The shape records...
        fillColor: Color;               // Had a color, and 
        case kind: ShapeType of    // Shape Data is either...
            CIRCLE_TYPE:    ( circ: Circle;         );
            RECTANGLE_TYPE: ( rect: Rectangle;      );
            //Todo: Add Ellipse and Triangle
    end;
    
    Drawing = record                             // A Drawing records has
        Shapes: array[0..MAX_SHAPES-1] of shape; // a number of shapes
        Index: integer;                          // the index of the last shape
        SelectedShape: ShapeType;                // the selected shape type
    end;

// ======================
// = Drawing Procedures =
// ======================

// Draw the shapes in the menu
procedure DrawShapesMenu(selectedShape: ShapeType);
begin
    // Draw the "toolbar" area
    FillRectangle(ColorLightGrey,0,0,60,600);
    
    // Draw the menu shapes.
    FillCircle(ColorGreen, MENU_CIRCLE);
    FillRectangle(ColorGreen, MENU_RECT);
    //Todo: Add Ellipse and Triangle
    
    // Redraw the selected shape
    case(SelectedShape) of
        RECTANGLE_TYPE: DrawRectangle(ColorBlack, MENU_RECT);
        CIRCLE_TYPE:    DrawCircle(ColorBlack, MENU_CIRCLE);
        //Todo: Add Ellipse and Triangle
    end;
end;